# Week 5 Homework

In this homework we'll put what we learned about Spark in practice.

For this homework we will be using the FHVHV 2021-06 data found here. [FHVHV Data](https://github.com/DataTalksClub/nyc-tlc-data/releases/download/fhvhv/fhvhv_tripdata_2021-06.csv.gz )

### Question 1: 

**Install Spark and PySpark** 

- Install Spark
- Run PySpark
- Create a local spark session
- Execute spark.version.

What's the output?
- `3.3.2`
- 2.1.4
- 1.2.3
- 5.4
</br></br>

In [1]:
import pyspark
from pyspark.sql import SparkSession

In [3]:
spark = SparkSession.builder \
        .master("local[*]") \
        .appName('test') \
        .getOrCreate()

In [4]:
spark.version

'3.0.3'

### Question 2: 

**HVFHW June 2021**

Read it with Spark using the same schema as we did in the lessons.</br> 
We will use this dataset for all the remaining questions.</br>
Repartition it to 12 partitions and save it to parquet.</br>
What is the average size of the Parquet (ending with .parquet extension) Files that were created (in MB)? Select the answer which most closely matches.</br>


- 2MB
- `24MB`
- 100MB
- 250MB
</br></br>

In [9]:
!wget https://github.com/DataTalksClub/nyc-tlc-data/releases/download/fhvhv/fhvhv_tripdata_2021-06.csv.gz -P data/raw/fhvhv/

--2023-03-06 14:04:50--  https://github.com/DataTalksClub/nyc-tlc-data/releases/download/fhvhv/fhvhv_tripdata_2021-06.csv.gz
Resolving github.com (github.com)... 20.205.243.166, ::ffff:20.205.243.166
Connecting to github.com (github.com)|20.205.243.166|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/4564ad9e-a6da-4923-ad6f-35ff02446a51?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20230306%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20230306T140448Z&X-Amz-Expires=300&X-Amz-Signature=152fd55584566aca9f803f848258dc271a44d5ec4b7f52336f5fbe20772bdf75&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=513814948&response-content-disposition=attachment%3B%20filename%3Dfhvhv_tripdata_2021-06.csv.gz&response-content-type=application%2Foctet-stream [following]
--2023-03-06 14:04:51--  https://objects.githubusercontent.com/github-production-release-asse

In [18]:
import pandas as pd

df = pd.read_csv("./data/raw/fhvhv/fhvhv_tripdata_2021-06.csv.gz", nrows=100)

In [19]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 7 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   dispatching_base_num    100 non-null    object
 1   pickup_datetime         100 non-null    object
 2   dropoff_datetime        100 non-null    object
 3   PULocationID            100 non-null    int64 
 4   DOLocationID            100 non-null    int64 
 5   SR_Flag                 100 non-null    object
 6   Affiliated_base_number  66 non-null     object
dtypes: int64(2), object(5)
memory usage: 5.6+ KB


In [ ]:
from pyspark.sql import types

In [ ]:
schema = types.StructType([
    types.StructField('dispatching_base_num', types.StringType(), True),
    types.StructField('pickup_datetime', types.TimestampType(), True),
    types.StructField('dropoff_datetime', types.TimestampType(), True),
    types.StructField('PULocationID', types.IntegerType(), True),
    types.StructField('DOLocationID', types.IntegerType(), True),
    types.StructField('SR_Flag', types.StringType(), True),
    types.StructField('Affiliated_base_number', types.StringType(), True),
])

In [ ]:
df = spark \
        .read \
        .option("header", "true") \
        .schema(schema) \
        .csv("./data/raw/fhvhv/fhvhv_tripdata_2021-06.csv.gz")

In [11]:
df.show()

+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+
|dispatching_base_num|    pickup_datetime|   dropoff_datetime|PULocationID|DOLocationID|SR_Flag|Affiliated_base_number|
+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+
|              B02764|2021-06-01 00:02:41|2021-06-01 00:07:46|         174|          18|      N|                B02764|
|              B02764|2021-06-01 00:16:16|2021-06-01 00:21:14|          32|         254|      N|                B02764|
|              B02764|2021-06-01 00:27:01|2021-06-01 00:42:11|         240|         127|      N|                B02764|
|              B02764|2021-06-01 00:46:08|2021-06-01 00:53:45|         127|         235|      N|                B02764|
|              B02510|2021-06-01 00:45:42|2021-06-01 01:03:33|         144|         146|      N|                  null|
|              B02510|2021-06-01 00:18:1

In [ ]:
df.repartition(12) \
        .write.parquet("./data/pq/fhvhv/", mode = "overwrite")    

In [5]:
spark_df = spark.read.parquet('./data/pq/fhvhv/*')

In [6]:
spark_df.show()

+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+
|dispatching_base_num|    pickup_datetime|   dropoff_datetime|PULocationID|DOLocationID|SR_Flag|Affiliated_base_number|
+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+
|              B02872|2021-06-12 05:01:33|2021-06-12 05:15:11|         220|          74|      N|                B02872|
|              B02879|2021-06-18 18:08:44|2021-06-18 18:18:35|          42|         152|      N|                B02879|
|              B02866|2021-06-30 17:15:50|2021-06-30 17:18:45|         123|         123|      N|                B02866|
|              B02617|2021-06-16 21:54:55|2021-06-16 22:05:43|          89|          89|      N|                B02617|
|              B02764|2021-06-06 21:07:12|2021-06-06 21:25:48|         248|         265|      N|                B02764|
|              B02510|2021-06-02 21:49:3

In [7]:
spark_df.count()

14961892

In [8]:
! wc -c ./data/pq/fhvhv/*

        0 ./data/pq/fhvhv/_SUCCESS
 24752738 ./data/pq/fhvhv/part-00000-eea3b2a5-d661-4c14-a3bf-0c29d0042746-c000.snappy.parquet
 24739695 ./data/pq/fhvhv/part-00001-eea3b2a5-d661-4c14-a3bf-0c29d0042746-c000.snappy.parquet
 24751808 ./data/pq/fhvhv/part-00002-eea3b2a5-d661-4c14-a3bf-0c29d0042746-c000.snappy.parquet
 24750821 ./data/pq/fhvhv/part-00003-eea3b2a5-d661-4c14-a3bf-0c29d0042746-c000.snappy.parquet
 24751984 ./data/pq/fhvhv/part-00004-eea3b2a5-d661-4c14-a3bf-0c29d0042746-c000.snappy.parquet
 24752334 ./data/pq/fhvhv/part-00005-eea3b2a5-d661-4c14-a3bf-0c29d0042746-c000.snappy.parquet
 24727599 ./data/pq/fhvhv/part-00006-eea3b2a5-d661-4c14-a3bf-0c29d0042746-c000.snappy.parquet
 24750886 ./data/pq/fhvhv/part-00007-eea3b2a5-d661-4c14-a3bf-0c29d0042746-c000.snappy.parquet
 24751822 ./data/pq/fhvhv/part-00008-eea3b2a5-d661-4c14-a3bf-0c29d0042746-c000.snappy.parquet
 24751996 ./data/pq/fhvhv/part-00009-eea3b2a5-d661-4c14-a3bf-0c29d0042746-c000.snappy.parquet
 24752936 ./data/pq/fhvhv

24000000 = 24 MB

### Question 3: 

**Count records**  

How many taxi trips were there on June 15?</br></br>
Consider only trips that started on June 15.</br>

- 308,164
- 12,856
- `452,470`
- 50,982
</br></br>

In [9]:
 spark_df.createOrReplaceTempView('fhvhv')

In [18]:
spark.sql("""
SELECT
    count(*)
FROM
    fhvhv
WHERE date(pickup_datetime) = "2021-06-15"
""").show()

+--------+
|count(1)|
+--------+
|  452470|
+--------+



### Question 4: 

**Longest trip for each day**  

Now calculate the duration for each trip.</br>
How long was the longest trip in Hours?</br>

- `66.87 Hours`
- 243.44 Hours
- 7.68 Hours
- 3.32 Hours
</br></br>

In [38]:
from pyspark.sql.functions import col

In [73]:
spark_df = spark_df.withColumn('duration_hr', col("dropoff_datetime").cast("long") - col("pickup_datetime").cast("long"))

In [74]:
spark_df.show()

+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+-----------+
|dispatching_base_num|    pickup_datetime|   dropoff_datetime|PULocationID|DOLocationID|SR_Flag|Affiliated_base_number|duration_hr|
+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+-----------+
|              B02872|2021-06-12 05:01:33|2021-06-12 05:15:11|         220|          74|      N|                B02872|        818|
|              B02879|2021-06-18 18:08:44|2021-06-18 18:18:35|          42|         152|      N|                B02879|        591|
|              B02866|2021-06-30 17:15:50|2021-06-30 17:18:45|         123|         123|      N|                B02866|        175|
|              B02617|2021-06-16 21:54:55|2021-06-16 22:05:43|          89|          89|      N|                B02617|        648|
|              B02764|2021-06-06 21:07:12|2021-06-06 21:25:48|         248| 

In [75]:
 spark_df.createOrReplaceTempView('fhvhv')

In [76]:
spark.sql("""
SELECT 
       to_date(dropoff_datetime) as date,
       duration_hr/(60*60)
  FROM fhvhv
 ORDER BY 2 DESC;
""").show()

+----------+---------------------------------------------------------+
|      date|(CAST(duration_hr AS DOUBLE) / CAST((60 * 60) AS DOUBLE))|
+----------+---------------------------------------------------------+
|2021-06-28|                                         66.8788888888889|
|2021-06-23|                                       25.549722222222222|
|2021-06-28|                                       19.980833333333333|
|2021-06-27|                                       18.197222222222223|
|2021-06-24|                                       16.466944444444444|
|2021-06-24|                                       14.268888888888888|
|2021-06-25|                                       13.909722222222221|
|2021-06-05|                                                    11.67|
|2021-06-27|                                       11.365833333333333|
|2021-06-21|                                       10.984444444444444|
|2021-06-01|                                                  10.2675|
|2021-

### Question 5: 

**User Interface**

 Spark’s User Interface which shows application's dashboard runs on which local port?</br>

- 80
- 443
- `4040`
- 8080
</br></br>


![Alt text](../images/spark_ui.JPG)

### Question 6: 

**Most frequent pickup location zone**

Load the zone lookup data into a temp view in Spark</br>
[Zone Data](https://github.com/DataTalksClub/nyc-tlc-data/releases/download/misc/taxi_zone_lookup.csv)</br>

Using the zone lookup data and the fhvhv June 2021 data, what is the name of the most frequent pickup location zone?</br>

- East Chelsea
- Astoria
- Union Sq
- `Crown Heights North`
</br></br>

In [77]:
!wget https://github.com/DataTalksClub/nyc-tlc-data/releases/download/misc/taxi_zone_lookup.csv .

--2023-03-07 17:38:48--  https://github.com/DataTalksClub/nyc-tlc-data/releases/download/misc/taxi_zone_lookup.csv
Resolving github.com (github.com)... 20.205.243.166
Connecting to github.com (github.com)|20.205.243.166|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/5a2cc2f5-b4cd-4584-9c62-a6ea97ed0e6a?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20230307%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20230307T103849Z&X-Amz-Expires=300&X-Amz-Signature=7f1d67e7eb9e817f98cc409ebce25b1114a89936d51e9a3692423a8ae7b569d8&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=513814948&response-content-disposition=attachment%3B%20filename%3Dtaxi_zone_lookup.csv&response-content-type=application%2Foctet-stream [following]
--2023-03-07 17:38:49--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/5a2cc2f5-b4cd-4584-9c62

In [79]:
zone_spdf = spark \
        .read \
        .option("header", "true") \
        .csv("taxi_zone_lookup.csv")

In [81]:
 zone_spdf.createOrReplaceTempView('zone_spdf')

In [92]:
spark.sql("""
SELECT 
      *
  FROM fhvhv
  LIMIT 1
""").show()

+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+-----------+
|dispatching_base_num|    pickup_datetime|   dropoff_datetime|PULocationID|DOLocationID|SR_Flag|Affiliated_base_number|duration_hr|
+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+-----------+
|              B02872|2021-06-12 05:01:33|2021-06-12 05:15:11|         220|          74|      N|                B02872|        818|
+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+-----------+



In [93]:
spark.sql("""
SELECT 
      *
  FROM zone_spdf
  LIMIT 1
""").show()

+----------+-------+--------------+------------+
|LocationID|Borough|          Zone|service_zone|
+----------+-------+--------------+------------+
|         1|    EWR|Newark Airport|         EWR|
+----------+-------+--------------+------------+



In [98]:
spark.sql("""
SELECT 
      zone,
      count(*)
  FROM fhvhv
  LEFT JOIN zone_spdf
  ON fhvhv.PULocationID = zone_spdf.LocationID
  GROUP BY 1
  ORDER BY 2 DESC
  LIMIT 1
""").show()

+-------------------+--------+
|               zone|count(1)|
+-------------------+--------+
|Crown Heights North|  231279|
+-------------------+--------+

